gradle-baseline-java/src/main/java/com/palantir/baseline/plugins/BaselineTesting.java

```diff
@@ -64,88 +56,52 @@ public void apply(Project project) {
 
-            project.getExtensions()
-                    .getByType(JavaPluginExtension.class)
-                    .getSourceSets()
-                    .configureEach(sourceSet -> {
-                        getTestTaskForSourceSet(project, sourceSet).ifPresent(testTask -> {
-                            testTask.dependsOn(checkJUnitDependencies);
-                        });
-
-                        ifHasResolvedCompileDependenciesMatching(
-                                project,
-                                sourceSet,
-                                BaselineTesting::requiresJunitPlatform,
-                                () -> fixSourceSet(project, sourceSet));
+            testingExtension
+                    .getSuites()
+                    .named(JvmTestSuitePlugin.DEFAULT_TEST_SUITE_NAME, JvmTestSuite.class, testSuite -> {
+                        testSuite.useJUnitJupiter();
                     });
-        });
-    }
-
-    private void fixSourceSet(Project project, SourceSet ss) {
-        Optional<Test> maybeTestTask = getTestTaskForSourceSet(project, ss);
-        if (!maybeTestTask.isPresent()) {
-            log.warn("Detected 'org:junit.jupiter:junit-jupiter', but unable to find test task");
-            return;
-        }
-        log.info(
-                "Detected 'org:junit.jupiter:junit-jupiter', enabling useJUnitPlatform() on {}",
-                maybeTestTask.get().getName());
-        enableJunit5ForTestTask(maybeTestTask.get());
-    }
-
-    public static Optional<Test> getTestTaskForSourceSet(Project proj, SourceSet ss) {
-        String testTaskName = ss.getTaskName(null, "test");
 
-        Task task1 = proj.getTasks().findByName(testTaskName);
-        if (task1 instanceof Test) {
-            return Optional.of((Test) task1);
-        }
-
-        // unbroken dome does this
-        Task task2 = proj.getTasks().findByName(ss.getName());
-        if (task2 instanceof Test) {
-            return Optional.of((Test) task2);
-        }
-        return Optional.empty();
-    }
+            javaPluginExtension.getSourceSets().configureEach(sourceSet -> {
+                TaskCollection<Test> testTasks = project.getTasks()
+                        .withType(Test.class)
+                        .matching(task -> task.getName().equals(sourceSet.getName()));
+
+                // We must eagerly create the test task in order to add the junit-platform-launcher dependency.
+                // Otherwise the dependencies will already have been resolved by the time we create the test task.
+                testTasks.all(task -> {
+                    // See https://github.com/gradle/gradle/pull/21919
+                    if (GradleVersion.current().compareTo(GRADLE_8) < 0) {
+                        project.getDependencies()
+                                .add(sourceSet.getRuntimeOnlyConfigurationName(), JUNIT_PLATFORM_LAUNCHER);
+                    }
```

Comment:
Gradle didn't start automatically adding this dependency until Gradle 8, so for uses on Gradle 7 we replicate that behavior here.

https://github.com/gradle/gradle/pull/21919

Once we drop support for Gradle 7, this should be removed.